### DISCUSSION TITLE
use in-flight preload requests instead of cancel+refetch

### DISCUSSION THREAD

**Discussion by mpgon** [remix-run/remix#5776] [Proposals]
I am noticing that I'm finding this scenario somewhat often: 
```
-> hover on Link prefetch="intent" 
-> data preload starts 
-> click/navigate to link while preload is incomplete 
-> in-flight preload request is (sometimes) cancelled 
-> new data request for the loader is fired
```
I wonder if there would be a way to catch that there is data request in-flight, and wait for that to complete, instead of doing the behaviour above.
I tried to illustrate this in the video attached. The UX difference can be noticeable. 

https://user-images.githubusercontent.com/3008451/224569131-d6d99ebb-5eb6-47ff-a143-b1abb213b109.mov

In a tangential DX discussion, I wonder if there would also be a way to automatically reuse the preloaded data if the user went directly from hover to click (without having to setup manual cache control headers for this specific case). Seems like an easy to miss optimisation and one that would be expected to be there. There are fine grained cases that probably should be checked like if the user un-hovers the link then hovers and clicks, the stale data shouldn't be used. But I'm not seeing a situation where a Link prefetch is setup, hovered and clicked, and a loader is fired but not intended to be used.

**Comment by mpgon** [comment]
I guess for the first question it's not really possible [per spec](https://developer.mozilla.org/en-US/docs/Web/HTTP/Link_prefetching_FAQ#what_happens_if_i_click_on_a_link_while_something_is_being_prefetched)

**Comment by sergiodxa** [comment]
The prefetch is owned by the browser because Remix uses the `<link rel="prefetch" />` tag to do it, and it then uses `fetch` to get the data, so it can't know if the browser already completed the prefetch, the benefit of this approach is that the browser could choose to ignore the prefetch if it's busy doing something else or based on user preferences (like save data).

**Comment by mpgon** [comment]
Thanks, yeah I see, I checked how Link implements prefetch after posting the original question (sorry 😄), if this ever exists it needs to be implemented at the browser level.

Something else I would really love to be able to do with Remix but probably would also need to be implemented at the browser level would be a way to invalidate a HTTP cache entry manually. Having to implement a server side cache for the simple case of "keep this route cached in the browser for 1 minute unless an action is performed" is a bit sad but the only way around I found was using react-query or apollo (the ones I tested) in those routes instead of loaders, which works but has other tradeoffs in the data management flow in Remix (which I'm aware). Couldn't Remix add a cache bursting for fetching the loaders e.g. with [{ cache: "reload" }](https://developer.mozilla.org/en-US/docs/Web/API/Request/cache) when they are triggered by an Action? Wouldn't this always make sense and align with the philosophy of refetch all loaders on actions (excluding shouldRevalidate)?
